File: 백준/Silver/15900. 나무 탈출/나무 탈출.cc
#include <bits/stdc++.h>

using namespace std;

int turn = 0;
int dep[500001] = {0, };
vector<int> edges[500001];
bool visited[500001] = {false, };

void dfs(int cur) {
    visited[cur] = true;
    bool isLeaf = true;
    for(auto &next: edges[cur]) {
        if(!visited[next]) {
            isLeaf = false;
            dep[next] = dep[cur]+1;
            dfs(next);
        }
    }
    
    if(isLeaf) turn += dep[cur];
}

int main()
{
    int N; cin >> N;
    for(int i = 0; i < N-1; i++) {
        int a, b; cin >> a >> b;
        edges[a].push_back(b);
        edges[b].push_back(a);
    }
    
    dfs(1);
    
    cout << (turn%2 ? "Yes" : "No");
}

File: 백준/Silver/15900. 나무 탈출/나무 탈출.py
import sys
import math
from collections import deque

#sys.stdin = open('input.txt')
input = sys.stdin.readline

N = int(input())
dep = [0 for i in range(0, N + 1)]
edges = [[] for i in range(0, N + 1)]
visited = [False for i in range(0, N + 1)]
turn = 0


def bfs(start):
    global turn
    queue = deque()
    queue.append(start)
    visited[start] = True

    while len(queue) > 0:
        cur = queue.popleft()
        is_leaf = True

        for next in edges[cur]:
            if not visited[next]:
                is_leaf = False
                dep[next] = dep[cur] + 1
                queue.append(next)
                visited[next] = True
        if is_leaf:
            turn += dep[cur]


for i in range(N - 1):
    a, b = map(int, input().split())
    edges[a].append(b)
    edges[b].append(a)

bfs(1)

if turn % 2:
    print('Yes')
else:
    print('No')